server/pulp/server/db/querysets.py

```diff
@@ -1,10 +1,44 @@
-from mongoengine.queryset import DoesNotExist, QuerySet
+from mongoengine.queryset import DoesNotExist, QuerySetNoCache
 from pymongo import ASCENDING
 
 from pulp.server import exceptions as pulp_exceptions
 
 
-class CriteriaQuerySet(QuerySet):
+class CustomQuerySet(QuerySetNoCache):
+    """
+    All custom QuerySet classes should inherit from this class rather than QuerySet
+    or QuerySetNoCache. There are two reasons for this. Firstly, all QuerySets should
+    not cache their query results by default, so QuerySetNoCache must be inherited.
+    Secondly, QuerySetNoCache's implementation of the `cache` instance method
+    explicitly returns an instance of QuerySet which effectively removes any
+    customized behavior for that QuerySet.
+    """
+
+    def cache(self):
+        """
+        Currently, there is no reason to have a caching version of any of our
+        custom QuerySet classes. If that ever changes, this method should be
+        implemented to return a caching version of the custom QuerySet. This
+        implementation exists in order to avoid bugs resulting from the
+        unexpected default behavior of `cache`.
+
+        As of mongoengine 0.10, the default implementation of ``cache`` cannot
+        be used as it explicitly creates a QuerySet object, thus removing any
+        custom behavior of the custom QuerySet.
+        """
+        raise NotImplementedError()
+
+    def __len__(self):
```

Comment:
On Tue, Dec 08, 2015 at 01:00:58PM -0800, Michael Hrivnak wrote:
> I'm in favor of sticking with one way to do it. ``.count()`` is the generally accepted and supported way to get the count from the db, so let's not confuse the matter. I don't think ``len()`` is sufficiently more convenient, readable, or easier to use to justify a divergence of behavior from the rest of the library.
> 
> "There should be one-- and preferably only one --obvious way to do it."

I don't believe that Python "mantra" is about this problem. The
mongoengine implementation of __len__ is problematic, and even has some
controversy surrounding it upstream. This implementation was not about
making it so that Pulp devs *will* use it, it was more about making sure
that we don't balloon our RAM in the event that someone does use it.

Upstream is arguably the place that shouldn't have provided more than
one way to do it, so for us as a downstream user that ship has sailed.

I'm a +1 for overriding the upstream __len__(). I don't believe there is
a reasonable chance that we would ever want their behavior for this
function and this implementation makes sure that we keep our RAM usage
down in the event that someone accidentally does this.

An alternative could be to override it and have it raise a
NotImplemented("Please use the count() method instead."). I think that
would make both of us happy (you because then we're enforcing that
there's only one way, and me because we won't have any future
"accidents" in using len()). What do you think about that?

-- 
Randy Barlow
irc:   bowlofeggs
phone: +1-919-890-8150